solution/q2/util/main.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/sqs"
+)
+
+// listS3Keys retrieves the list of object keys from the given S3 bucket.
+func listS3Keys(s3Client *s3.S3, bucket string, startKey string, keyChannel chan<- string) {
+	input := &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+	}
+	foundStartKey := false
+
+	err := s3Client.ListObjectsV2Pages(input,
+		func(page *s3.ListObjectsV2Output, lastPage bool) bool {
+			for _, obj := range page.Contents {
+				if startKey != "" && !foundStartKey && *obj.Key == startKey {
+					foundStartKey = true
```

Comment:
Is there a more efficient way to restart a failed listing here, than having to page through all the previously-processed objects until we find the one we want?

Reply:
An improvement I can see here is to use the "full list with batch-like processing" I described above. So instead of iterating over first part of `ListObjectsV2Input` output, we'd restart directly from an "offset" that is known to be the place where previous execution was interrupted.